### PR TITLE
Expose `rz_core_analysis_name`

### PR DESCRIPTION
**Pull request by b1llow**
**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

* `core_analysis_name` -> exposed {`rz_core_analysis_rename`, `rz_core_analysis_name`}

**Test plan**

...

**Closing issues**

...

### CODE REVIEW

File: librz/include/rz_core.h [hunk 1]
```diff
@@ -756,6 +756,20 @@ RZ_API void rz_core_analysis_esil_init_mem_del(RZ_NONNULL RzCore *core, RZ_NULLA
 RZ_API void rz_core_analysis_esil_init_regs(RZ_NONNULL RzCore *core);
 
 /* canalysis.c */
+typedef enum rz_analysis_name_type {
+	VAR = 0,
+	FUNCTION,
+	FLAG,
+	ADDRESS,
```
Comment (wargio) [inline librz/include/rz_core.h]: these needs a better name.
what about `RZ_ANALYSIS_NAME_TYPE_XXX` ?

**Review by wargio — CHANGES_REQUESTED** [review]

File: librz/core/canalysis.c [hunk 2]
```diff
@@ -7126,3 +7126,107 @@ RZ_API st64 rz_core_analysis_calls_count(RZ_NONNULL RzCore *core) {
...
 }
+
+static const char *RzAnalysisNameTypeStrs[] = {
+	"var",
+	"function",
+	"flag",
+	"address",
+};
+
+RZ_IPI const char *rz_analysis_name_type_to_str(RzAnalysisNameType typ) {
```
Comment (wargio) [inline librz/core/canalysis.c]: maybe also expose this one. do not forget ownership.

File: librz/core/canalysis.c [hunk 3]
```diff
@@ -7126,3 +7126,107 @@ RZ_API st64 rz_core_analysis_calls_count(RZ_NONNULL RzCore *core) {
...
+	case FLAG:
+	case ADDRESS:
+		return RzAnalysisNameTypeStrs[typ];
+	default:
+		rz_warn_if_reached();
+		return NULL;
+	}
+}
+
+RZ_API void rz_analysis_name_free(RzAnalysisName *p) {
```
Comment (wargio) [inline librz/core/canalysis.c]: RZ_NULLABLE.

**Review by XVilka — APPROVED** [review]

File: librz/include/rz_core.h [hunk 4]
```diff
@@ -756,6 +756,20 @@ RZ_API void rz_core_analysis_esil_init_mem_del(RZ_NONNULL RzCore *core, RZ_NULLA
...
+	RZ_ANALYSIS_NAME_TYPE_FLAG,
+	RZ_ANALYSIS_NAME_TYPE_ADDRESS,
+} RzAnalysisNameType;
+
+typedef struct rz_analysis_name {
+	char *name;
+	char *realname;
+	RzAnalysisNameType type;
+	ut64 offset;
+} RzAnalysisName;
```
Comment (ret2libc) [inline librz/include/rz_core.h]:
```suggestion
} RzCoreAnalysisName;
```

File: librz/include/rz_core.h [hunk 5]
```diff
@@ -756,6 +756,20 @@ RZ_API void rz_core_analysis_esil_init_mem_del(RZ_NONNULL RzCore *core, RZ_NULLA
 RZ_API void rz_core_analysis_esil_init_regs(RZ_NONNULL RzCore *core);
 
 /* canalysis.c */
+typedef enum rz_analysis_name_type {
+	RZ_ANALYSIS_NAME_TYPE_VAR = 0,
+	RZ_ANALYSIS_NAME_TYPE_FUNCTION,
+	RZ_ANALYSIS_NAME_TYPE_FLAG,
+	RZ_ANALYSIS_NAME_TYPE_ADDRESS,
+} RzAnalysisNameType;
```
Comment (ret2libc) [inline librz/include/rz_core.h]:
```suggestion
} RzCoreAnalysisNameType;
```

File: librz/include/rz_core.h [hunk 6]
```diff
@@ -806,6 +820,11 @@ RZ_API st64 rz_core_analysis_coverage_count(RZ_NONNULL RzCore *core);
 RZ_API st64 rz_core_analysis_code_count(RZ_NONNULL RzCore *core);
 RZ_API st64 rz_core_analysis_calls_count(RZ_NONNULL RzCore *core);
 
+RZ_API RZ_BORROW const char *rz_analysis_name_type_to_str(RzAnalysisNameType typ);
+RZ_API void rz_analysis_name_free(RZ_NULLABLE RzAnalysisName *p);
```
Comment (ret2libc) [inline librz/include/rz_core.h]:
```suggestion
RZ_API RZ_BORROW const char *rz_core_analysis_name_type_to_str(RzAnalysisNameType typ);
RZ_API void rz_core_analysis_name_free(RZ_NULLABLE RzAnalysisName *p);
```

File: librz/core/cmd/cmd_analysis.c [hunk 7]
```diff
@@ -373,98 +373,50 @@ static void list_vars(RzCore *core, RzAnalysisFunction *fcn, PJ *pj, int type, c
...
-				pj_o(pj);
-				pj_ks(pj, "name", var->name);
-				pj_ks(pj, "type", "var");
-				pj_kn(pj, "offset", tgt_addr);
-				pj_end(pj);
-			} else {
-				rz_cons_println(var->name);
-			}
+static bool core_analysis_name_print(RzCore *core, const char *name, PJ *pj) {
+	if (name) {
```
Comment (ret2libc) [inline librz/core/cmd/cmd_analysis.c]: Please check if the name argument is provided in the command handler and directly call `rz_core_analysis_rename` there if necessary. This is a `_print` function thus it should just print.

File: librz/core/cmd/cmd_analysis.c [hunk 8]
```diff
@@ -373,98 +373,50 @@ static void list_vars(RzCore *core, RzAnalysisFunction *fcn, PJ *pj, int type, c
...
-			if (pj) {
-				pj_o(pj);
-				pj_ks(pj, "name", var->name);
-				pj_ks(pj, "type", "var");
-				pj_kn(pj, "offset", tgt_addr);
-				pj_end(pj);
-			} else {
-				rz_cons_println(var->name);
-			}
+static bool core_analysis_name_print(RzCore *core, const char *name, PJ *pj) {
```
Comment (ret2libc) [inline librz/core/cmd/cmd_analysis.c]: just pass `state` instead of PJ.

File: librz/core/cmd/cmd_analysis.c [hunk 9]
```diff
@@ -373,98 +373,50 @@ static void list_vars(RzCore *core, RzAnalysisFunction *fcn, PJ *pj, int type, c
...
-	RzAnalysisOp op;
-	ut64 tgt_addr = UT64_MAX;
-
-	if (pj) {
-		pj_a(pj);
+#define PJ_KS(pj, key, value) \
+	{ \
+		const char *value_tmp = (value); \
+		if (RZ_STR_ISNOTEMPTY(value_tmp)) { \
+			pj_ks(pj, key, value_tmp); \
```
Comment (ret2libc) [inline librz/core/cmd/cmd_analysis.c:380]: @thestr4ng3r @XVilka i wonder whether we should just check in the `pj` API whether the value is present or not.. WDYT? For pj_kn maybe we can have a separate API to avoid assuming UT64_MAX as "empty value".

File: librz/core/canalysis.c [hunk 10]
```diff
@@ -7126,3 +7126,110 @@ RZ_API st64 rz_core_analysis_calls_count(RZ_NONNULL RzCore *core) {
...
+		}
+	}
+	return false;
+}
+
+/**
+ * \brief Get information on whatever var/flag/function is used at core->offset
+ * \return RzAnalysisName
+ */
+RZ_API RZ_OWN RzAnalysisName *rz_core_analysis_name(RZ_NONNULL RzCore *core) {
```
Comment (ret2libc) [inline librz/core/canalysis.c]: don't assume core->offset, pass an address as argument. This way the API can be more easily used even elsewhere by just passing a different address, without the need to switch core->offset value.

File: librz/core/canalysis.c [hunk 11]
```diff
@@ -7126,3 +7126,110 @@ RZ_API st64 rz_core_analysis_calls_count(RZ_NONNULL RzCore *core) {
...
+	if (!p) {
+		return NULL;
+	}
+
+	RzAnalysisOp op;
+	ut64 tgt_addr = UT64_MAX;
+
+	rz_analysis_op(core->analysis, &op, core->offset,
+		core->block, 32, RZ_ANALYSIS_OP_MASK_BASIC);
+	RzAnalysisVar *var = rz_analysis_get_used_function_var(core->analysis, op.addr);
```
Comment (ret2libc) [inline librz/core/canalysis.c:7219]: From rz_analysis.h this function is deprecated:
```
// There could be multiple vars used in multiple functions. Use rz_analysis_get_functions_in()+rz_analysis_function_get_vars_used_at() instead.
```

Comment (ret2libc) [reply librz/core/canalysis.c:7219]: However... I wonder whether we should just rename this to `get_used_function_var_first` or something like that. @thestr4ng3r ?

File: librz/core/canalysis.c [hunk 12]
```diff
@@ -7126,3 +7126,110 @@ RZ_API st64 rz_core_analysis_calls_count(RZ_NONNULL RzCore *core) {
...
+	free(p->name);
+	free(p->realname);
+	free(p);
+}
+
+/**
+ * \brief Rename whatever var/flag/function is used at core->offset to \p name
+ * \return success?
+ */
+RZ_API bool rz_core_analysis_rename(RZ_NONNULL RzCore *core, RZ_NONNULL const char *name) {
```
Comment (ret2libc) [inline librz/core/canalysis.c]: add address arg here as well

File: librz/core/canalysis.c [hunk 13]
```diff
@@ -7126,3 +7126,110 @@ RZ_API st64 rz_core_analysis_calls_count(RZ_NONNULL RzCore *core) {
...
+	RzAnalysisVar *var = rz_analysis_get_used_function_var(core->analysis, op.addr);
+	tgt_addr = op.jump != UT64_MAX ? op.jump : op.ptr;
+	rz_analysis_op_fini(&op);
+	if (var) {
+		return rz_analysis_var_rename(var, name, true);
+	} else if (tgt_addr != UT64_MAX) {
+		RzAnalysisFunction *fcn = rz_analysis_get_function_at(core->analysis, tgt_addr);
+		RzFlagItem *f = rz_flag_get_i(core->flags, tgt_addr);
+		if (fcn) {
+			return rz_analysis_function_rename(fcn, name);
```
Comment (ret2libc) [inline librz/core/canalysis.c]: I wonder whether we should use `rz_core_analysis_function_rename` instead.

**Review by ret2libc — CHANGES_REQUESTED** [review]

File: librz/core/canalysis.c [hunk 14]
```diff
@@ -7126,3 +7126,135 @@ RZ_API st64 rz_core_analysis_calls_count(RZ_NONNULL RzCore *core) {
...
+ * \return success?
+ */
+RZ_API bool rz_core_analysis_rename(RZ_NONNULL RzCore *core, RZ_NONNULL const char *name, ut64 addr) {
+	rz_return_val_if_fail(core && core->analysis && RZ_STR_ISNOTEMPTY(name), false);
+
+	ut64 oldoff = core->offset;
+	if (addr != core->offset) {
+		rz_core_seek(core, addr, true);
+	}
+
```
Comment (ret2libc) [inline librz/core/canalysis.c]: Why is this required?

File: librz/core/canalysis.c [hunk 15]
```diff
@@ -7126,3 +7126,135 @@ RZ_API st64 rz_core_analysis_calls_count(RZ_NONNULL RzCore *core) {
...
+	if (!p) {
+		return NULL;
+	}
+	p->offset = UT64_MAX;
+
+	ut64 oldoff = core->offset;
+	if (addr != core->offset) {
+		rz_core_seek(core, addr, true);
+	}
+
```
Comment (ret2libc) [inline librz/core/canalysis.c]: Again, why this?

File: librz/core/canalysis.c (hunk 13 above)
Comment (b1llow) [reply librz/core/canalysis.c]: Maybe

File: librz/core/canalysis.c [hunk 16]
```diff
@@ -7126,3 +7126,135 @@ RZ_API st64 rz_core_analysis_calls_count(RZ_NONNULL RzCore *core) {
...
+	rz_analysis_op(core->analysis, &op, core->offset,
+		core->block, 32, RZ_ANALYSIS_OP_MASK_BASIC);
+	RzAnalysisVar *var = rz_analysis_get_used_function_var(core->analysis, op.addr);
+	ut64 tgt_addr = op.jump != UT64_MAX ? op.jump : op.ptr;
+	rz_analysis_op_fini(&op);
+
+	if (var) {
+		p->name = strdup(var->name);
+		p->type = RZ_CORE_ANALYSIS_NAME_TYPE_VAR;
+		p->offset = tgt_addr;
```
Comment (ret2libc) [inline librz/core/canalysis.c]: Is this tgt_addr or op.addr?

File: librz/core/canalysis.c [hunk 17]
```diff
@@ -7126,3 +7126,135 @@ RZ_API st64 rz_core_analysis_calls_count(RZ_NONNULL RzCore *core) {
...
+		} else if (f) {
+			p->type = RZ_CORE_ANALYSIS_NAME_TYPE_FLAG;
+			p->name = strdup(f->name);
+			p->realname = strdup(f->realname);
+			p->offset = tgt_addr;
+		} else {
+			p->type = RZ_CORE_ANALYSIS_NAME_TYPE_ADDRESS;
+			p->offset = tgt_addr;
+		}
+	}
```
Comment (ret2libc) [inline librz/core/canalysis.c]: put an else here and do the core_analysis_name_free+return NULL.

File: librz/core/canalysis.c (hunk 14 above)
Comment (b1llow) [reply librz/core/canalysis.c]: I see that we always use `core->block` as a global buffer, then only the data of `core->block[core->offset..core->offset+core->blocksize]` is valid, so even if it is given an `addr` argument but it does not guarantee that the data at the `addr` is valid.

Comment (b1llow) [reply librz/core/canalysis.c]: Forget it when I didn't say

File: librz/core/canalysis.c (hunk 16 above)
Comment (b1llow) [reply librz/core/canalysis.c]: op.addr
i fix it.

File: librz/core/cmd/cmd_analysis.c (hunk 9 above)
Comment (XVilka) [reply librz/core/cmd/cmd_analysis.c:380]: Maybe

File: librz/core/canalysis.c [hunk 18]
```diff
@@ -7126,3 +7126,124 @@ RZ_API st64 rz_core_analysis_calls_count(RZ_NONNULL RzCore *core) {
...
+}
+
+/**
+ * \brief Rename whatever var/flag/function is used at \p addr to \p name
+ * \return success?
+ */
+RZ_API bool rz_core_analysis_rename(RZ_NONNULL RzCore *core, RZ_NONNULL const char *name, ut64 addr) {
+	rz_return_val_if_fail(core && core->analysis && RZ_STR_ISNOTEMPTY(name), false);
+
+	ut8 buf[32];
```
Comment (ret2libc) [inline librz/core/canalysis.c]: maybe we should use `rz_analysis_archinfo` + `RZ_ANALYSIS_ARCHINFO_MAX_OP_SIZE` to determine the maxsize of the buffer here? I'm not sure how it is done around the codebase.. I bet it's quite inconsistent. @wargio do you know?

**Review by ret2libc — APPROVED** [review]
LGTM, good work!

File: librz/core/canalysis.c (hunk 18 above)
Comment (wargio) [reply librz/core/canalysis.c]: yes technically you should do that, but normally 128 bytes are always enough, 32 not so sure...

File: librz/core/canalysis.c [hunk 19]
```diff
@@ -7126,3 +7126,124 @@ RZ_API st64 rz_core_analysis_calls_count(RZ_NONNULL RzCore *core) {
...
+	if (!rz_io_read_at(core->io, addr, buf, 32)) {
```
Comment (wargio) [inline librz/core/canalysis.c]:
```suggestion
	if (!rz_io_read_at(core->io, addr, buf, sizeof(buf))) {
```

File: librz/core/canalysis.c [hunk 20]
```diff
@@ -7126,3 +7126,124 @@ RZ_API st64 rz_core_analysis_calls_count(RZ_NONNULL RzCore *core) {
...
+		return false;
+	}
+
+	RzAnalysisOp op;
+	rz_analysis_op(core->analysis, &op, core->offset,
+		buf, 32, RZ_ANALYSIS_OP_MASK_BASIC);
```
Comment (wargio) [inline librz/core/canalysis.c]:
```suggestion
		buf, sizeof(buf), RZ_ANALYSIS_OP_MASK_BASIC);
```

File: librz/core/canalysis.c [hunk 21]
```diff
@@ -7126,3 +7126,124 @@ RZ_API st64 rz_core_analysis_calls_count(RZ_NONNULL RzCore *core) {
...
+
+/**
+ * \brief Get information on whatever var/flag/function is used at \p addr
+ * \return RzAnalysisName
+ */
+RZ_API RZ_OWN RzCoreAnalysisName *rz_core_analysis_name(RZ_NONNULL RzCore *core, ut64 addr) {
+	rz_return_val_if_fail(core && core->analysis, NULL);
+
+	ut8 buf[32];
+	if (!rz_io_read_at(core->io, addr, buf, 32)) {
```
Comment (wargio) [inline librz/core/canalysis.c]:
```suggestion
	if (!rz_io_read_at(core->io, addr, buf, sizeof(buf))) {
```

File: librz/core/canalysis.c [hunk 22]
```diff
@@ -7126,3 +7126,124 @@ RZ_API st64 rz_core_analysis_calls_count(RZ_NONNULL RzCore *core) {
...
+	}
+
+	RzCoreAnalysisName *p = RZ_NEW0(RzCoreAnalysisName);
+	if (!p) {
+		return NULL;
+	}
+
+	RzAnalysisOp op;
+	rz_analysis_op(core->analysis, &op, core->offset,
+		buf, 32, RZ_ANALYSIS_OP_MASK_BASIC);
```
Comment (wargio) [inline librz/core/canalysis.c]:
```suggestion
		buf, sizeof(buf), RZ_ANALYSIS_OP_MASK_BASIC);
```

**Review by wargio — APPROVED** [review]
Greate job! just address the `sizeof(buf)`

File: librz/core/canalysis.c (hunk 18 above)
Comment (ret2libc) [reply librz/core/canalysis.c]: @imbillow please use 128 for now... I guess we should at least define a macro in rz_analysis.h in the future for that magic number.

Comment (b1llow) [reply librz/core/canalysis.c]: OK

**Review by wargio — APPROVED** [review]

**Review by XVilka — APPROVED** [review]
